packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -458,12 +486,29 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
         document.removeEventListener('keydown', onKeyDown, true);
       }
 
-      if (restoreFocus && nodeToRestore && isElementInScope(document.activeElement, scopeRef.current)) {
-        requestAnimationFrame(() => {
-          if (document.body.contains(nodeToRestore)) {
-            focusElement(nodeToRestore);
-          }
-        });
+      if (nodeToRestore) {
+        let index = nodeToRestoreArray.indexOf(nodeToRestore);
+
+        if (restoreFocus && isElementInScope(document.activeElement, scopeRef.current)) {
+          requestAnimationFrame(() => {
+            if (document.body.contains(nodeToRestore) && nodeToRestore !== document.body) {
+              focusElement(nodeToRestore);
+            } else {
+              while (index > 0) {
+                index--;
+                let node = nodeToRestoreArray[index];
+                if (!document.body.contains(node) || node === document.body) {
+                  removeFromNodeToRestoreArray(node);
+                } else {
+                  focusElement(node);
+                  break;
+                }
+              }
+            }
+          });
+        } else if (!document.body.contains(nodeToRestore)) {
+          removeFromNodeToRestoreArray(nodeToRestore);
+        }
       }
     };
   }, [scopeRef, restoreFocus, contain]);
```

Comment:
As an aside and probably out of scope of this PR since it is preexisting, but doesn't this `return` fire not only as a cleanup but also if any of these dep array values change? That means we could unintentionally call `focusElement` and move focus out of the active scope if `contain` were to update values